### DISCUSSION TITLE
rc.17 中使用ref 获取元素 页面卡死

### DISCUSSION THREAD

**Discussion by Nokecy** [alibaba/formily#2383] [Q&A 1.x] [answered]
https://codesandbox.io/s/vigorous-feynman-84vql?file=/App.tsx

注释 ref: barCodeRef 页面可以显示 ，使用原生的<Input ref={barCodeRef} />  也可以正常操作，是我的姿势不对吗  @janryWang  大佬

**Comment by janryWang** [comment, marked as answer]
不推荐这样使用，ref引用给formily会被转换成observable对象，这种操作dom的，建议直接用原生API

**Reply by Nokecy** [reply]
大佬 大概怎么玩～～ 能不能写个demo @janryWang

**Comment by baoxiangyang** [comment]
我也有遇到情况， 我的自定义组件可能会对外暴露方法， 但是我外部的ref 通过observable自后就丢失了。
我组件要对外暴露方法的话，应该如何处理呢? @janryWang

**Comment by janryWang** [comment]
cosnt useFormilyRef = <T>(current:T):{current:T}=>useMemo(()=>markRaw({current}),[])

**Comment by csrigogogo** [comment]
补充一下 @formily/reactive 的 markRaw 实现 是通过 Symbol 做标记 , 如果出现两个版本的@formily/reactive 的包, 应该是要保持两个包版本一致
